### PR TITLE
feat: Add drift information to the CRP status

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
### Description of your changes

Add drift information to the CRP status

Fixes #

I have:

- [x] Run `make reviewable` to ensure this PR is ready for review.

### CODE REVIEW

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 1]
```diff
@@ -657,53 +657,78 @@ const (
...
 
+// DriftedResourcePlacement contains the drifted resource details.
+type DriftedResourcePlacement struct {
+	// The resource that has drifted.
+	// +kubebuilder:validation:Required
+	ResourceIdentifier `json:",inline"`
+
+	// The ObservedGeneration of the resource on the target cluster that caused this drift.
+	// +kubebuilder:validation:Required
+	ObservedGeneration int64 `json:"observedGeneration"`
```
Comment (michaelawyu) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:717]: Hi Ryan! Just a quick question: this assumes that we will be switching to our own condition implementations to relay this piece of information back, right?

Comment (michaelawyu) [reply apis/placement/v1beta1/clusterresourceplacement_types.go:717]: Also, considering that `observedGeneration` is kind of a well-known name in the K8s community, do we want to be more specific that it's about the generation on the target cluster? This is just a nit though, of course.

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 2]
```diff
@@ -657,53 +657,78 @@ const (
...
+
+	// The details of difference between the resource on the target cluster and the resource to be applied from the hub.
+	// We will also emit an event with the difference details.
+	// +kubebuilder:validation:maxLength=8192
+	// +kubebuilder:validation:Optional
+	DriftDetail string `json:"driftDetail,omitempty"`
```
Comment (michaelawyu) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:723]: TBH I am a bit worried about the size limit as well. Should we mention directly that we might be truncating this information as necessary and the user is expected to check on their own for large diffs?

File: apis/placement/v1beta1/clusterresourceplacement_types.go (hunk 1 above)
Comment (michaelawyu) [reply apis/placement/v1beta1/clusterresourceplacement_types.go:717]: Just a random thought, but a `driftedSince` field could also be useful, just to pinpoint the cause

**Review by michaelawyu — COMMENTED** [review]
Just a few nits; otherwise LGTM

File: apis/placement/v1beta1/clusterresourceplacement_types.go [hunk 3]
```diff
@@ -657,53 +657,78 @@ const (
...
 	Condition metav1.Condition `json:"condition"`
 }
 
+// DriftedResourcePlacement contains the drifted resource details.
+type DriftedResourcePlacement struct {
+	// The resource that has drifted.
+	// +kubebuilder:validation:Required
+	ResourceIdentifier `json:",inline"`
+
+	// The ObservedGeneration of the resource on the target cluster that caused this drift.
```
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:715]:
```suggestion
	// The ObservedGeneration of the resource on the target cluster that contains this drift.
```

File: apis/placement/v1beta1/clusterresourceplacement_types.go (hunk 1 above)
Comment (circy9) [reply apis/placement/v1beta1/clusterresourceplacement_types.go:717]: Add observedAt to denote when the observation is made?

File: apis/placement/v1beta1/clusterresourceplacement_types.go (hunk 2 above)
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:723]: Use ObservedDifferences?

Drift is a noun to describe the action of A becoming more and more different from B.

File: apis/placement/v1beta1/clusterresourceplacement_types.go (hunk 1 above)
Comment (circy9) [inline apis/placement/v1beta1/clusterresourceplacement_types.go:717]: ObservedGenerationInTargetCluster?

**Review by circy9 — APPROVED** [review]

File: apis/placement/v1beta1/clusterresourceplacement_types.go (hunk 1 above)
Comment (circy9) [reply apis/placement/v1beta1/clusterresourceplacement_types.go]: done

**Pull request merged**